### PR TITLE
[FIXED JENKINS-39555] Don't do the actions initialization by helper method

### PR DESCRIPTION
**Pull request by svanoort**
Root cause for [JENKINS-39555](https://issues.jenkins-ci.org/browse/JENKINS-39555)  -- fixes regression introduced in https://github.com/jenkinsci/jenkins/pull/2608.  

Resolves this in testing against workflow-api-plugin. 

Checklist:

- [x] Verify basic fix resolves issues
- [x] Check that overrides handle mutation correctly
- [x] Unit test to ensure this catastrophe isn't replicated
- [x] Verify threading behavior

### CODE REVIEW

**Comment by svanoort** [comment]
@jenkinsci/code-reviewers  @reviewbybees

File: core/src/main/java/hudson/model/Actionable.java [hunk 1]
```diff
@@ -77,24 +76,12 @@
      */
     @Deprecated
```
Comment (abayer) [inline core/src/main/java/hudson/model/Actionable.java]: Should this still be `Deprecated`?

**Review by abayer — APPROVED** [review]
:bee: modulo my question about whether `getActions()` should still be `Deprecated`.

**Comment by oleg-nenashev** [comment]
LGTM 🐝

**Comment by oleg-nenashev** [comment]
@reviewbybees done

File: core/src/main/java/hudson/model/Actionable.java (hunk 1 above)
Comment (abayer) [inline core/src/main/java/hudson/model/Actionable.java]: Fair enough!

**Comment by stephenc** [comment]
:bee:

File: core/src/main/java/hudson/model/Actionable.java [hunk 2]
```diff
@@ -77,24 +76,12 @@
...
-     * {@link #replaceAction(Action)}, {@link #removeAction(Action)}, etc.
-     * @return the CopyOnWriteArrayList of persisted actions.
-     */
-    private CopyOnWriteArrayList<Action> getOrCreateActions() {
-        if(actions == null) {
-            synchronized (this) {
-                if(actions == null) {
-                    actions = new CopyOnWriteArrayList<Action>();
-                }
+        synchronized (this) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Actionable.java]: Simpler to just mark the method `synchronized`.

File: core/src/main/java/hudson/model/Actionable.java [hunk 3]
```diff
@@ -137,16 +124,14 @@
      * Adds a new action.
      * Note: calls to {@link #getAllActions()} that happen before calls to this method may not see the update.
      * <strong>Note: this method will always modify the actions</strong>
-     *
-     * The default implementation is mostly equivalent to the call chain {@code getActions().add(a)}.
```
Comment (jglick) [inline core/src/main/java/hudson/model/Actionable.java]: Better to restore that note and just delete the word `mostly`!

File: test/src/test/java/hudson/model/ActionableTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package hudson.model;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.jvnet.hudson.test.Issue;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ActionableTest {
```
Comment (jglick) [inline test/src/test/java/hudson/model/ActionableTest.java]: This is not using `JenkinsRule` so it should rather be in `core/src/test/java/hudson/model/ActionableTest.java`.

File: test/src/test/java/hudson/model/ActionableTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Issue("JENKINS-39555")
+    @Test
+    public void testExtensionOverrides() throws Exception {
+        ActionableOverride myOverridden = new ActionableOverride();
+        InvisibleAction invis = new InvisibleAction() {
+        };
+        myOverridden.addAction(invis);
+        Assert.assertArrayEquals(new Object[]{invis}, myOverridden.specialActions.toArray());
```
Comment (jglick) [inline test/src/test/java/hudson/model/ActionableTest.java]:
```java
assertEquals(Collections.singletonList(invis), myOverridden.specialActions);
```

File: test/src/test/java/hudson/model/ActionableTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        Assert.assertArrayEquals(new Object[]{invis}, myOverridden.getActions().toArray());
+
+        myOverridden.getActions().remove(invis);
```
Comment (jglick) [inline test/src/test/java/hudson/model/ActionableTest.java]: What about `removeAction`?

**Review by jglick — APPROVED** [review]
:bee: though it would be better for the tests to be folded into the existing `ActionableTest`.

File: core/src/main/java/hudson/model/Actionable.java (hunk 3 above)
Comment (svanoort) [inline core/src/main/java/hudson/model/Actionable.java]: Doesn't really add anything then IMO - especially when it's 3 LoC

**Comment by abayer** [comment]
re-:bee:

**Comment by christ66** [comment]
:bee:

**Comment by batmat** [comment]
:+1:

File: core/src/test/java/hudson/model/ActionableTest.java [hunk 7]
```diff
@@ -48,6 +53,56 @@ public void replaceAction() {
...
+
+    @SuppressWarnings("deprecation")
+    @Issue("JENKINS-39555")
+    @Test
+    public void testExtensionOverrides() throws Exception {
+        ActionableOverride myOverridden = new ActionableOverride();
+        InvisibleAction invis = new InvisibleAction() {
+        };
+        myOverridden.addAction(invis);
+        Assert.assertArrayEquals(new Object[]{invis}, myOverridden.specialActions.toArray());
```
Comment (jglick) [inline core/src/test/java/hudson/model/ActionableTest.java]: Same suggestion as before.

**Review by jglick — APPROVED** [review]
re:bee:

File: core/src/main/java/hudson/model/Actionable.java [hunk 8]
```diff
@@ -137,16 +124,14 @@
...
      */
     @SuppressWarnings({"ConstantConditions","deprecation"})
     @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
     public void addAction(@Nonnull Action a) {
         if(a==null) {
             throw new IllegalArgumentException("Action must be non-null");
         }
-        getOrCreateActions().add(a);
+        getActions().add(a);
```
Comment (jglick) [inline core/src/main/java/hudson/model/Actionable.java]: [JENKINS-42969](https://issues.jenkins-ci.org/browse/JENKINS-42969)

**Pull request merged**